src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPolymorphismOptions.cs

```diff
@@ -0,0 +1,107 @@
+using System.Reflection;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    /// <summary>
+    /// Defines polymorphic configuration for a specified base type.
+    /// </summary>
+    public class JsonPolymorphismOptions
+    {
+        private DerivedTypeList? _derivedTypes;
+        private bool _ignoreUnrecognizedTypeDiscriminators;
+        private JsonUnknownDerivedTypeHandling _unknownDerivedTypeHandling;
+        private string? _typeDiscriminatorPropertyName;
+
+        /// <summary>
+        /// Gets the list of derived types supported in the current polymorphic type configuration.
+        /// </summary>
+        public IList<JsonDerivedType> DerivedTypes => _derivedTypes ??= new(this);
+
+        /// <summary>
+        /// When set to <see langword="true"/>, instructs the serializer to ignore any
+        /// unrecognized type discriminator id's and reverts to the contract of the base type.
+        /// Otherwise, it will fail the deserialization.
+        /// </summary>
+        public bool IgnoreUnrecognizedTypeDiscriminators
+        {
+            get => _ignoreUnrecognizedTypeDiscriminators;
+            set
+            {
+                VerifyMutable();
+                _ignoreUnrecognizedTypeDiscriminators = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the behavior when serializing an undeclared derived runtime type.
+        /// </summary>
+        public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling
+        {
+            get => _unknownDerivedTypeHandling;
+            set
+            {
+                VerifyMutable();
+                _unknownDerivedTypeHandling = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a custom type discriminator property name for the polymorhic type.
+        /// Uses the default '$type' property name if left unset.
+        /// </summary>
+        [AllowNull]
+        public string TypeDiscriminatorPropertyName
+        {
+            get => _typeDiscriminatorPropertyName ?? JsonSerializer.TypePropertyName;
+            set
+            {
+                VerifyMutable();
+                _typeDiscriminatorPropertyName = value;
+            }
+        }
+
+        private void VerifyMutable() => DeclaringTypeInfo?.VerifyMutable();
+
+        internal JsonTypeInfo? DeclaringTypeInfo { get; set; }
+
+        private sealed class DerivedTypeList : ConfigurationList<JsonDerivedType>
+        {
+            private readonly JsonPolymorphismOptions _parent;
+
+            public DerivedTypeList(JsonPolymorphismOptions parent)
+            {
+                _parent = parent;
+            }
+
+            protected override bool IsLockedInstance => _parent.DeclaringTypeInfo?.IsConfigured == true;
+            protected override void VerifyMutable() => _parent.DeclaringTypeInfo?.VerifyMutable();
+        }
+
+        internal static JsonPolymorphismOptions? CreateFromAttributeDeclarations(Type baseType)
```

Comment:
unrelated to this PR: we should use similar pattern for ReflectionJsonTypeInfo and other similar places, it makes it much cleaner than randomly reading attributes in random places